com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/flowable/FlowableFacade.java

```diff
@@ -226,42 +241,128 @@ public void trigger(String executionId, Map<String, Object> variables) {
-                }
-                LOGGER.warn(format(Messages.RETRYING_PROCESS_ABORT, processInstanceId));
+                throwExceptionInCaseOfTimeout(processInstanceId, overallAbortDeadline, e);
+            } catch (NotFoundException e) {
+                LOGGER.error(e.getMessage(), e);
+                return;
+            }
+        }
+    }
+
+    private void abortProcessSafely(String processInstanceId, String deleteReason) {
+        // Get all active leaf executions before aborting all running processes
+        List<Execution> allActiveExecutionsWithoutChildren = getAllActiveExecutionsWithoutChildren(processInstanceId);
+        setAbortProcessVariableInContext(processInstanceId);
+        waitAllActiveExecutionsToFinish(processInstanceId, allActiveExecutionsWithoutChildren);
+        suspendProcessIfNotSuspended(processInstanceId);
+        abortProcessInstance(processInstanceId, deleteReason);
+    }
+
+    private void setAbortProcessVariableInContext(String processInstanceId) {
+        LOGGER.debug(format(Messages.SETTING_VARIABLE, Constants.PROCESS_ABORTED, Boolean.TRUE));
+        processEngine.getRuntimeService()
+                     .setVariable(processInstanceId, Constants.PROCESS_ABORTED, Boolean.TRUE);
+    }
+
+    private void suspendProcessIfNotSuspended(String processInstanceId) {
+        if (isProcessSuspended(processInstanceId)) {
+            LOGGER.info(format(PROCESS_0_HAS_ALREADY_BEEN_SUSPENDED, processInstanceId));
+            return;
+        }
+        suspendProcessInstance(processInstanceId);
+    }
+
+    private boolean isProcessSuspended(String processInstanceId) {
+        Optional<Execution> rootProcessExecution = getRootExecution(processInstanceId);
+        return rootProcessExecution.isPresent() && rootProcessExecution.get()
+                                                                        .isSuspended();
+    }
+
+    private void waitAllActiveExecutionsToFinish(String processInstanceId, List<Execution> allActiveExecutionsWithoutChildren) {
+        List<Execution> activeExecutionsWithoutChildren = allActiveExecutionsWithoutChildren;
+        long allSubprocessesFinishedDeadline = System.currentTimeMillis() + DEFAULT_ABORT_WAIT_TIMEOUT_MS;
+        while (true) {
+            List<Execution> newActiveExecutionsWithoutChildren = getNewExecutionsWithoutChildren(processInstanceId,
+                                                                                                 activeExecutionsWithoutChildren);
+            LOGGER.info(format(NEW_ACTIVE_EXECUTIONS_WITHOUT_CHILDREN_0_FOR_PROCESS_1, newActiveExecutionsWithoutChildren,
+                               processInstanceId));
+            activeExecutionsWithoutChildren = ListUtils.union(activeExecutionsWithoutChildren, newActiveExecutionsWithoutChildren);
+            activeExecutionsWithoutChildren = removeExecutionsIfTheyHaveChildren(processInstanceId, activeExecutionsWithoutChildren);
```

Comment:
Do you need this method ? The above two methods getNewExecutionsWithoutChildren and getAllActiveExecutionsWithoutChildren assume that you are will not have executios which have children.